Add RSA tests with actual RSA keypair implementation

Limess opened the pull request:
Replaces the previous test which was entirely based around stubs.

Generates an RSA key pair based on the s3o implementation: 'https://github.com/Financial-Times/s3o/blob/992be3a122006210013da64aa2ceee65e03a479d/src/main/java/com/ft/s3o/util/KeyGenerator.java#L19'.
Signs a request as username-host as per s3o and verifies that validate returns true for matching keys.

Commit:
Add RSA tests with actual RSA keypair implementation

Replaces the previous test which was entirely based around stubs.

Generates an RSA key pair based on the s3o implementation: 'https://github.com/Financial-Times/s3o/blob/992be3a122006210013da64aa2ceee65e03a479d/src/main/java/com/ft/s3o/util/KeyGenerator.java#L19'.
Signs a request as username-host as per s3o and verifies that validate returns true for matching keys.

wheresrhys commented on validate.js:
```diff
@@ -11,7 +11,7 @@ const NodeRSA = require('node-rsa');
  */
 module.exports = function (s3oPublicKey) {
-	return function (key, token) {
+	return function (key, signedToken) {
```
is there also a less generic name than 'key' that can be used here

Limess replied:
I was trying to think of one... It's a generic piece of data, could be anything - even `key` is a bit much for this token to know.

I've just changed this whole signature to match https://nodejs.org/api/crypto.html#crypto_verify_verify_object_signature_signatureencoding

wheresrhys commented on test/validate.spec.js:
```diff
@@ -45,31 +46,46 @@ describe('lib/validate.js', () => {
+		});
+
+		it('returns true if supplied key validates against the public key', () => {
+			const token = sign('charlie.briggs:lantern.ft.com', key)
```
Why si the token coon separated, but the token in the previous test hyphen separated?

Limess replied:
Oops, last minute change

wheresrhys commented on test/validate.spec.js:
```diff
@@ -45,31 +46,46 @@ describe('lib/validate.js', () => {
 
-		it('returns true if supplied key validates against public key', () => {
-			keyFactory.returns('test-public-key');
+		it('returns false if supplied key does not validates against public key', () => {
```
```suggestion
		it('returns false if supplied key does not validate against public key', () => {
```

wheresrhys commented on test/validate.spec.js:
```diff
@@ -45,31 +46,46 @@ describe('lib/validate.js', () => {
-		it('returns false if publicKey is falsy', () => {
-			keyFactory.returns(false);
+		['', false, undefined, null].forEach((value) => {
+			it(`returns false if publicKey is ${value}`, () => {
```
```suggestion
			it(`returns false if publicKey is "${value}"`, () => {
```

wheresrhys commented on test/validate.spec.js:
```diff
@@ -1,42 +1,43 @@
 
 describe('lib/validate.js', () => {
+	let publicKey
+	let key
```
is this variable used anywhere?

Limess replied:
yup. `publicKey` is used to setup stubs. `key` is used to in each `sign` call.

Renamed to `keyPair` for clarity.

Commit:
Review comments

Fix typos, make naming clearer

Commit:
Rename key to object, token to signature as per node crypto docs

See 'https://nodejs.org/api/crypto.html#crypto_verify_verify_object_signature_signatureencoding'

This is a generic method for verifying some string against a signature, we don't know what format the object is in this function

Commit:
Update README to trigger rebuild

Trying to get Snyk to validate
